spec/01-unit/01-db/01-schema/01-schema_spec.lua

```diff
@@ -2989,4 +3015,249 @@ describe("schema", function()
+            end,
+          },
+        },
+      }
+      local entity = { name = "test1" }
+      local input = { name = nil }
+
+      local TestEntities = Schema.new(test_schema)
+      local transformed_entity, _ = TestEntities:transform(entity, input)
+
+      assert.truthy(transformed_entity)
+      assert.equal("test1", transformed_entity.name)
+    end)
+
+
+    it("transforms fields with multiple transformations", function()
+      local test_schema = {
+        name = "test",
+        fields = {
+          {
+            name = {
+              type = "string"
+            },
+          },
+        },
+        transformations = {
+          {
+            input = { "name" },
+            func = function(name)
+              return { name = "How are you " .. name }
+            end,
+          },
+          {
+            input = { "name" },
+            func = function(name)
+              return { name = name .. "?" }
+            end,
+          },
+        },
+      }
+
+      local entity = { name = "Bob" }
+
+      local TestEntities = Schema.new(test_schema)
+      local transformed_entity, _ = TestEntities:transform(entity)
+
+      assert.truthy(transformed_entity)
+      assert.equal("How are you Bob?", transformed_entity.name)
+    end)
+
+    it("returns error if transformation returns an error", function()
+      local test_schema = {
+        name = "test",
+        fields = {
+          {
+            name = {
+              type = "string"
+            },
+          },
+        },
+        transformations = {
+          {
+            input = { "name" },
+            func = function(name)
+              return nil, "unable to transform name"
+            end,
+          },
+        },
+      }
+      local entity = { name = "test1" }
+
+      local TestEntities = Schema.new(test_schema)
+      local transformed_entity, err = TestEntities:transform(entity)
+
+      assert.falsy(transformed_entity)
+      assert.equal("transformation failed: unable to transform name", err)
+    end)
+
+    it("skips transformation if needs are not fulfilled", function()
+      local test_schema = {
+        name = "test",
+        fields = {
+          {
+            name = {
+              type = "string"
+            },
+            age = {
+              type = "integer"
+            },
+          },
+        },
+        transformations = {
+          {
+            input = { "name" },
+            needs = { "age" },
+            func = function(name)
+              return { name = name:upper() }
+            end,
+          },
+        },
+      }
+      local entity = { name = "test1" }
+
+      local TestEntities = Schema.new(test_schema)
+      local transformed_entity, _ = TestEntities:transform(entity)
+
+      assert.truthy(transformed_entity)
+      assert.equal("test1", transformed_entity.name)
+    end)
+
+
+    it("transforms fields with needs given to function", function()
```

Comment:
I am unclear as to why we need `needs`. It seems this specific case could be handled inside the transformation itself with an `if`:

```
        transformations = {
          {
            input = { "name", "age" },
            func = function(name, age)
              if age then
                return { name = name .. " " .. age, age = age }
              else
                return { name = name }
              end
            end,
          },
        },
```

Reply:
Yes, indeed we could have:
```
transformation = function(context, entity, original_input)
  return { <new-entity> }
end
```

Just a general description:
1. `transformation.input` adds `mutually_required` check over the fieds defined in `transformation.input`, kinda an implicit `entity check` for `schema:validate` that does validation against the original (user) input, the normal entity validations work on entity that could have been merged with rbw entity, or which foreign keys where resolved, or autofields filled.
2. `transformation.input` is also used as a deciding factor with `schema:transform` to detect if this transformation needs to be executed (e.g. all input fields where given in original (user) input). Kinda like second time check but this time used for triggering the transformation and not for adding implicit validation.
3. `transformation.needs` are the extra parameters that the function needs as a function arguments that may or may not be on the entity (not the original user input). If those fields are not specified, it means the transformation is skipped. Kinda like soft `transformation.input`. It never adds validation, it is just about skipping transformation or executing it. The needs are added at the end of the arguments list when transformation is called.

Reply:
@hishamhm gave a better description below. I will update the code to match that.